InvenTree/machine/machine_type.py

```diff
@@ -0,0 +1,91 @@
+from typing import TYPE_CHECKING, Dict
+
+# Import only for typechecking, otherwise this throws cyclic import errors
+if TYPE_CHECKING:
+    from common.models import SettingsKeyType
+    from machine.models import Machine
+else:  # pragma: no cover
+    class Machine:
+        pass
+
+    class SettingsKeyType:
+        pass
+
+
+class BaseDriver:
+    """Base class for machine drivers
+
+    Attributes:
+        SLUG: Slug string for identifying a machine
+        NAME: User friendly name for displaying
+        DESCRIPTION: Description of what this driver does (default: "")
+    """
+
+    SLUG: str
+    NAME: str
+    DESCRIPTION: str
+
+    MACHINE_SETTINGS: Dict[str, SettingsKeyType]
+
+    @classmethod
+    def validate(cls):
+        def attribute_missing(key):
+            return not hasattr(cls, key) or getattr(cls, key) == ""
+
+        def override_missing(base_implementation):
+            return base_implementation == getattr(cls, base_implementation.__name__, None)
+
+        missing_attributes = list(filter(attribute_missing, ["SLUG", "NAME", "DESCRIPTION"]))
+        missing_overrides = list(filter(override_missing, getattr(cls, "requires_override", [])))
+
+        errors = []
+
+        if len(missing_attributes) > 0:
+            errors.append(f"did not provide the following attributes: {', '.join(missing_attributes)}")
+        if len(missing_overrides) > 0:
+            errors.append(f"did not override the required attributes: {', '.join(map(lambda attr: attr.__name__, missing_overrides))}")
+
+        if len(errors) > 0:
+            raise NotImplementedError(f"The driver '{cls}' " + " and ".join(errors))
```

Comment:
Nice approach; Could we factor this out in a base that is used both by BaseDriver and BaseMachineType? We have similar things for most mixins too.

Reply:
@matmair Where should that mixin be placed? Using `InvenTree/mixins.py` doesn't work, because in that class `Inventree.helpers` are imported that import models which throws an `AppRegistryNotReady` issue if I import my `ClassValidationMixin` in this file.

Reply:
It is a class (but not model) mixin so probably in InvenTree/helpers - I am a bit surprised that models are imported on top level. Maybe move it into a separate file "registry_helpers" or something like that.

Reply:
Im not sure if this is long or term the best solution. Had this situation often that incant import helpers directly. Maybe we should refactor the helpers in a separate pr to import the models on function level? Also this ClassValidationMixin could be used for current plugins.

Reply:
@matmair Does importing the models at function level have any impact on performance? If so what about two helper files? One that contains model code and one that doesn't?

Reply:
Not really, there is a very very small difference. Split sounds like a good idea.

Reply:
Ref: #4927